### PR TITLE
[react-router] Add generic param for staticContext

### PR DESCRIPTION
**Pull request by pret-a-porter**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by pret-a-porter** [comment]
Code example for new definitions https://reacttraining.com/react-router/web/example/static-router

File: types/react-router/index.d.ts [hunk 1]
```diff
@@ -58,11 +59,11 @@ export interface RedirectProps {
 }
 export class Redirect extends React.Component<RedirectProps, any> { }
 
-export interface RouteComponentProps<P> {
-  match: match<P>;
-  location: H.Location;
+export interface RouteComponentProps<P, C extends {} = {}> {
```
Comment (huy-nguyen) [inline types/react-router/index.d.ts]: There's already a `StaticContext` interface defined in `test/examples-from-react-router-website/StaticRouter.tsx`. Is there a connection between the two? Should static context have `statusCode` as an optional property? If that's the same as what you have, can you integrate that interface into your PR?

Comment (pret-a-porter) [reply types/react-router/index.d.ts]: Done

**Comment by apieceofbart** [comment]
@pret-a-porter question: why wasn't `staticContext` optional and instead requires explicit `undefined` ?

**Comment by pret-a-porter** [comment]
@apieceofbart because `staticContext` field always defined in RouteComponentProps object. If field always defined, it is not optional field.

**Comment by huy-nguyen** [comment]
I agree with @apieceofbart. If the end-user doesn't have to provide `staticContext`, it should be marked as optional (`staticContext?: C`) and not as `staticContext: C | undefined` because the latter requires that the user explicitly specify `staticContext: undefined`, which can be quite annoying.

**Comment by johnnyreilly** [comment]
Is this safe to merge? I'm hesitant too as there's disagreement...

**Comment by apieceofbart** [comment]
@pret-a-porter I don't see it required here: https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/Route.js#L29

@johnnyreilly what do you mean? this has already been merged

**Comment by apieceofbart** [comment]
render method doesn't return `RouterComponentProps` - that doesn't make sense. Render returns ReactNode.

This line gets `staticContext` `const { history, route, staticContext } = this.context.router;`
Since it's defined as `staticContext: PropTypes.object` it can be omitted.

exactly as in your example above (which btw should be `b = { a: undefined }` )

**Comment by pret-a-porter** [comment]
Oh, sorry, my mistake about `RouterComponentProps`. Yes, I want use it in my component and describe `staticContext`.

**Comment by pret-a-porter** [comment]
Sorry for my mistakes.

```
export class RouteStatus extends React.PureComponent<IProps, {}> {
  public render() {
    return (
      <Route
        render={routerComponentProps => {
          console.log('props', routerComponentProps);

          if (routerComponentProps.staticContext) {
            routerComponentProps.staticContext.statusCode = this.props.statusCode;
          }

          return this.props.children;
        }}
      />
    );
  }
}
```

In this example object `routerComponentProps` always has field `staticContext`. Why it should be optional?

**Pull request merged**